Integration test cluster for Byron;Shelley

Anviking opened the pull request:
# Issue Number

#1920 

# Overview

- [x] Add support for Byron;Shelley integration test cluster
- [x] Adjusted configs to work with the new version of cardano-node
- [x] withCluster now takes three actions:
      - onByron
      - onFork (when in shelley. We can send money to the shelley faucets here)
      - onClusterStart (when the stake pools have fully started)
- [x] CLI faucets are now byron facuets, and are generated automatically. This is both easier and more elegant.
- [x] The underlying logic of withCluster was kept somewhat flexible for easier experimentation.
- [x] **Todo:** For the Shelley integration tests to work, we need to send funds to the shelley faucet addresses, as `initialFunds` doesn't work when entering shelley through a hard-fork.
- [x] **Todo** clean up faucet config/code, and make sure special wallets are included

# Comments


```haskell
$ stack test cardano-wallet-shelley:integration
```

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Add support for Byron;Shelley integration test cluster

- Adjusted configs to work with the new version of cardano-node
- withCluster now takes three actions:
  - onByron
  - onFork (when in shelley)
  - onClusterStart (when the stake pools have fully started)
- CLI faucets are now byron facuets, and are generated automatically.
This is both easier and more elegant.
- The underlying logic of withCluster was kept abstract for easier
experimentation.

Contains a few squashed commits from Rodney.

Anviking committed:
Print cartouche when waiting for hard-fork to finish

Anviking committed:
Fixup incorrect docs

Anviking committed:
wip: add working sendFaucetFundsTo

Anviking committed:
Remove shelley wallets from shelley genesis (now automatic)

Anviking committed:
wip: Add byron faucet funds

Anviking committed:
Fixup: Add NO_POOLS option, and remove dev functions

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -452,49 +549,76 @@ withBFTNode
+        let makeModeSpecific x = case mode of
+                 RunByronShelley _ -> byronKeys . shelleyKeys $ x
+                 RunByron -> byronKeys x
+                 RunShelley -> shelleyKeys x
```
Any particular reason to try supporting all three modes? Wouldn't it be easier to just support the Byron/Shelley ?

Anviking replied:
I hoped it could let us still run the integration tests using a pure shelley chain (as an option), but that didn't work. So yeah, we can remove it.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -452,49 +549,76 @@ withBFTNode
-            setEnv "CARDANO_NODE_SOCKET_PATH" socket
-            (rawTx, faucetPrv) <- prepareKeyRegistration tr dir
-            tx <- signTx tr dir rawTx [faucetPrv]
-            submitTx tr "pre-registered stake key" tx
```
This is important and shouldn't be removed. I get that it can only happen after the hard-fork, but I don't see it moved at the relevant place.

Anviking replied:
Yes, you're right.

I mentioned it in the PR description as a remaining todo:
> Todo clean up faucet config/code, and make sure special wallets are included

Edit: Ah, right, I only thought about the mnemonic. This also (should) exist, yes.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -452,49 +549,76 @@ withBFTNode
+withBFTNode
+    tr
+    baseDir
+    i
```
I don't understand why this index is added. This suggests there could be several BFT nodes, but it it's not the case.

Anviking replied:
Yes, not needed anymore 👍

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -965,7 +1217,16 @@ submitTx tr name signedTx = do
+
+-- | Submit a transaction through a running node.
+submitTxNoRetry :: Tracer IO ClusterLog -> FilePath -> IO ()
+submitTxNoRetry tr signedTx = do
```
What's the use for this function? Why would we want to drop the retry behavior?

Anviking replied:
🤔 Yeah, you're right.  `submitTx` works fine, and we should prefer it to this.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -362,13 +424,22 @@ newtype PoolConfig = PoolConfig
 -- registering pools. Passing `0` as a number of pool will simply start a single
 -- BFT node.
+--
+-- The callback actions are not guaranteed to use the same node.
```
This comment looks scary. It's quite important that we use the relay node for sending transactions, otherwise they might be rolled back and screw the entire test setup.

Anviking replied:
The existing pool setup are using the BFT node to setup the stake pools.

As the PR is now, the BFT would also be used to transfer funds to the shelley faucet addresses.

Are you sure that is a problem? I think the code is looks slightly nicer if we don't have to start the relay until the end, as we used to.

KtorZ replied:
It shouldn't be a problem indeed since there won't be any stake pools at the time the faucet funds are transferred to cause any rollback.

KtorZ committed:
some cosmetic changes while reviewing code

Anviking committed:
fixup Crypto.HD

Anviking committed:
Fixup: remove submitTxNoRetry

KtorZ committed:
remove multi-mode support (default to byron;shelley) and rename setup files for consistency

KtorZ committed:
replace hard-coded occurences of 'Secure Passphrase' with a named constant

And in particular, fix the inconsistency between fixture wallets and
empty wallets that have different passphrases!

Anviking committed:
Automate generation of shelley faucet funds

Anviking committed:
Re-add call to prepareKeyRegistration, fix mistake in addr

Anviking commented on lib/shelley/test/integration/Main.hs:
```diff
@@ -220,25 +226,42 @@ specWithServer (tr, tracers) = aroundAll withContext . after tearDown
+    afterFork dir _ = do
+        let encodeAddr = T.unpack . encodeAddress @'Mainnet
+        let addresses = map (first encodeAddr) shelleyIntegrationTestFunds
+        sendFaucetFundsTo stdoutTextTracer dir addresses
```
The setup is sprinkled across integration/Main.hs, Faucet.hs, and Launch.hs, which feels a bit weird. Maybe we can resolve that sometime.

Anviking commented on lib/shelley/test/data/cardano-node-shelley/shelley-genesis.yaml:
```diff
@@ -0,0 +1,63 @@
+# b) the `systemStart` is hard-coded here to please the parser but is replaced dynamically
+#    by the same code generating the final node configuration for integration.
+#
+# c) initialFunds are defined as a list of singleton, and not as a key-value map like in the
```
Hm.🤔 We don't really control the order when we fill the shelley faucets now. Maybe this might cause some failures @KtorZ .

Maybe we have to make sure to always keep funds for a sequential wallet in a single tx (but we may have as many as 200 addresses to fill for a single wallet).

KtorZ replied:
This will be a problem for big wallets only, which might explain why the scenario is currently failing. I thought initial funds in the shelley genesis were ignored though?

Anviking replied:
> I thought initial funds in the shelley genesis were ignored though?

Yes. 

(I came to think about it when reading the outdated documentation comment. Now updated.)

Anviking committed:
Adjust documentation in shelley-genesis

KtorZ committed:
add little correction to size estimation for Byron change addresses

  indeed, for the size estimation, we generate dummy change outputs that necessary are shelley addresses. Yet, Byron addresses have different sizes, and in particular, Byron are longer than shelley addresses (Icarus are however shorter). So, we apply a correction after the whole size calculation for each change output because generating dummy Byron addresses would be a real pain in the a**

KtorZ committed:
get rid of Shelley's fee estimator from test integration

  This has been heavily reliable and is actually duplicating logic compared to the API. So instead, we query directly the API for fee estimation within the scenario and rely on a single implementation of fee estimation. This commits break every test in Shelley that are using the fee estimator, they'll be updated in subsequent commits

KtorZ committed:
replace fee estimator from Scenario/API/Shelley/Transactions with API calls

  I also removed some scenarios that were completely redundant with others (testing that the fee estimation endpoint works correctly is intrinsically done in other scenarios now that are calling it for pretty much every transactions).

KtorZ committed:
fix 'slotRangeFromTimeRange' time interpreter query

The lifting was not really done wrong. We shouldn't return 'Nothing' if one of the given bound is 'Nothing'.

rvl committed:
Use genesis block parameters for waitForHardFork in tests
